### PR TITLE
[BREAK] Validate incoming message schema

### LINKED ISSUE #9878: [DOCS & BUG] API to upload files accepts "title_file_download" as String
@MartinSchoeler Our docs is showing an example on `chat.postMessage` that the property `title_file_download` is an String. That needs to be changed.

@MarcosSpessatto Can you change this API to validate that the properties are following the correct types? The issue right now is that `title_file_download` accepts any type of value. We need to ensure all types are validated.

### LINKED ISSUE #10511: [BUG] Some API/Method is allowing to have "short" field as String
### Description:

This is a feedback from an Android user:

screen shot 2018-04-19 at 11 09 43 am

I guess we need to make sure before creating an attachment, that this field is always boolean.

### PR DESCRIPTION
**Pull request by MarcosSpessatto**
We now verify the message object, when being sent, matches our expected schema. Should it not, we throw an error.

Closes #9878
Closes #10511

### CODE REVIEW

File: packages/rocketchat-api/server/v1/chat.js [hunk 1]
```diff
@@ -106,6 +106,67 @@ RocketChat.API.v1.addRoute('chat.pinMessage', { authRequired: true }, {
...
+					title: String,
+					value: String
+				}));
+			};
+
+			const validateAttachment = (attachment) => {
+				check(attachment, Match.ObjectIncluding({
+					color: Match.Maybe(String),
+					text: Match.Maybe(String),
+					ts: Match.Maybe(String),
```
Comment (rafaelks) [inline packages/rocketchat-api/server/v1/chat.js]: `ts` is a String? 🤔

File: packages/rocketchat-api/server/v1/chat.js [hunk 2]
```diff
@@ -106,6 +106,67 @@ RocketChat.API.v1.addRoute('chat.pinMessage', { authRequired: true }, {
...
+					attachment.fields.map(validateAttachmentsFields);
+				}
+			};
+
+			attachments.map(validateAttachment);
+		};
+
+		const validateBodyParams = (bodyParams) => {
+			check(bodyParams, Match.ObjectIncluding({
+				roomId: Match.Maybe(String),
```
Comment (rafaelks) [inline packages/rocketchat-api/server/v1/chat.js]: Isn't this one REQUIRED?

File: tests/end-to-end/api/05-chat.js [hunk 3]
```diff
@@ -32,7 +32,7 @@ describe('[Chat]', function() {
 					author_icon: 'https://avatars.githubusercontent.com/u/850391?v=3',
 					title: 'Attachment Example',
 					title_link: 'https://youtube.com',
-					title_link_download: 'https://rocket.chat/download',
+					title_link_download: true,
```
Comment (rafaelks) [inline tests/end-to-end/api/05-chat.js]: I think we can have more tests here, with lots of different variations and some fail case. Can you do that, please?

**Comment by MarcosSpessatto** [comment]
Can you see again @rafaelks ?

**Comment by rafaelks** [comment]
@MarcosSpessatto Looking great, thanks for tests. LGTM.

**Comment by rafaelks** [comment]
I approved, but I would that someone else also revised the JS code, please. @RocketChat/core

**Comment by MarcosSpessatto** [comment]
@rafaelks, can you take a look again please? I have added the same verification in `chat.sendMessage` too.

**Comment by rodrigok** [comment]
@graywolf336 should he add the validation to the `RocketChat.sendMessage` function https://github.com/RocketChat/Rocket.Chat/blob/develop/packages/rocketchat-lib/server/functions/sendMessage.js#L3 ?

And why the webhooks should be able to bypass the verification? I don't agree with that.

**Comment by graywolf336** [comment]
Yes.

No, I'm saying with the way this pull request currently is they would be able to but my request was to not allow them to. :)

File: packages/rocketchat-lib/server/functions/sendMessage.js [hunk 4]
```diff
@@ -1,10 +1,69 @@
...
+			alias: Match.Maybe(String),
+			emoji: Match.Maybe(String),
+			avatar: Match.Maybe(String),
+			attachments: Match.Maybe(Array)
+		}));
+
+		if (Array.isArray(message.attachments) && message.attachments.length) {
+			validateBodyAttachments(message.attachments);
+		}
+	} catch (error) {
```
Comment (graywolf336) [inline packages/rocketchat-lib/server/functions/sendMessage.js]: Why are we catching an error here to then throw the same error? Let's not do that. Either allow the error to be thrown or throw a customized error.

File: packages/rocketchat-lib/server/functions/sendMessage.js [hunk 5]
```diff
@@ -1,10 +1,69 @@
...
+		throw error;
+	}
+
+	if (message.ts == null) {
+		message.ts = new Date();
+	}
+	message.u = _.pick(user, ['_id', 'username', 'name']);
```
Comment (graywolf336) [inline packages/rocketchat-lib/server/functions/sendMessage.js]: Instead of relying on underscore for this, let's just do it ourselves. :)

File: packages/rocketchat-lib/server/functions/sendMessage.js [hunk 6]
```diff
@@ -1,10 +1,69 @@
...
+		}));
+
+		if (Array.isArray(message.attachments) && message.attachments.length) {
+			validateBodyAttachments(message.attachments);
+		}
+	} catch (error) {
+		throw error;
+	}
+
+	if (message.ts == null) {
```
Comment (graywolf336) [inline packages/rocketchat-lib/server/functions/sendMessage.js]: Can we change this to `if(!message.ts) {`?

File: packages/rocketchat-lib/server/functions/sendMessage.js [hunk 7]
```diff
@@ -1,10 +1,69 @@
 import _ from 'underscore';
 
+const validateBodyAttachments = (attachments) => {
```
Comment (graywolf336) [inline packages/rocketchat-lib/server/functions/sendMessage.js]: As we are now going to be validating the messages and attachments, do we have some place where we document the schema in our documentation?

File: packages/rocketchat-lib/server/functions/sendMessage.js [hunk 8]
```diff
@@ -1,10 +1,69 @@
...
+};
+
 RocketChat.sendMessage = function(user, message, room, upsert = false) {
 	if (!user || !message || !room._id) {
 		return false;
 	}
 
+	try {
+		check(message, Match.ObjectIncluding({
+			msg: Match.Maybe(String),
```
Comment (graywolf336) [inline packages/rocketchat-lib/server/functions/sendMessage.js]: Let's also check the `_id` field, since it can be provided.

File: packages/rocketchat-lib/server/methods/sendMessage.js [hunk 9]
```diff
@@ -10,6 +10,10 @@ Meteor.methods({
 			});
 		}
 
+		if (!message.rid) {
+			throw new Error('The \'rid\' param, must be provided');
```
Comment (graywolf336) [inline packages/rocketchat-lib/server/methods/sendMessage.js]: Maybe say `The 'rid' property on the message object is missing.`? Does that make more sense?

File: packages/rocketchat-lib/server/functions/sendMessage.js (hunk 7 above)
Comment (MarcosSpessatto) [reply packages/rocketchat-lib/server/functions/sendMessage.js]: It seens we don't have @graywolf336, only in the [endpoint](https://rocket.chat/docs/developer-guides/rest-api/chat/postmessage/) docs. :(

Comment (graywolf336) [reply packages/rocketchat-lib/server/functions/sendMessage.js]: Any recommendations for where to place it?

cc: @MartinSchoeler

Comment (graywolf336) [reply packages/rocketchat-lib/server/functions/sendMessage.js]: Hey @MartinSchoeler, any ideas?

**Comment by sampaiodiego** [comment]
@graywolf336 I agree.

**Comment by graywolf336** [comment]
Changed to [BREAK] prefix.

File: packages/rocketchat-lib/server/functions/sendMessage.js [hunk 10]
```diff
@@ -1,10 +1,66 @@
-import _ from 'underscore';
+const validateAttachmentsFields = attachmentFields => {
```
Comment (graywolf336) [inline packages/rocketchat-lib/server/functions/sendMessage.js]: @ggazzo why did you change the code style from parenthesis around the arguments to not?

Comment (ggazzo) [reply packages/rocketchat-lib/server/functions/sendMessage.js]: because we dont need `( )`in this case :)

**Comment by tbaehr** [comment]
You are breaking a lot of stuff (e.g. webhooks) by now enforcing title on attachments

+

Exception in callback of async function: Error: Match error: Missing key 'title'     at check (/opt/Rocket.Chat/programs/server/packages/check.js:68:17)     at validateAttachmentsFields (/opt/Rocket.Chat/programs/server/packages/rocketchat_lib.js:5283:3)   

This is not really a good way to do so!

**Comment by graywolf336** [comment]
@tbaehr Did you open an issue so we can get this fixed?

**Comment by tbaehr** [comment]
+	check(attachmentFields, Match.ObjectIncluding({
+		short: Match.Maybe(Boolean),
+		title: String,
+		value: String
+	}));

does not mean optional, but before it was not checked so it could be left out, didnt make a separate Bug

**Comment by graywolf336** [comment]
@tbaehr That is because those two properties are actually required when you are providing a field on the attachment. This is because of the way the mobile applications and everything are setup to display the attachment fields.
